counterparty-core/counterpartycore/lib/transaction_helper/transaction_inputs.py

```diff
@@ -0,0 +1,439 @@
+import binascii
+import decimal
+import hashlib
+import logging
+import threading
+
+import bitcoin as bitcoinlib
+import cachetools
+
+from counterpartycore.lib import backend, config, exceptions, script, util
+from counterpartycore.lib.transaction_helper import p2sh_serializer, transaction_outputs
+
+logger = logging.getLogger(config.LOGGER_NAME)
+D = decimal.Decimal
+
+MAX_INPUTS_SET = 100
+
+
+class BaseThreadSafeCache:
+    def __init__(self, *args, **kwargs):
+        # Note: reading is thread safe out of the box
+        self.lock = threading.Lock()
+        self.__cache = self.create_cache(*args, **kwargs)
+
+    def create_cache(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def get(self, key, default=None):
+        return self.__cache.get(key, default)
+
+    def pop(self, key, default=None):
+        with self.lock:
+            return self.__cache.pop(key, default)
+
+    def delete(self, key):
+        with self.lock:
+            try:
+                del self.__cache[key]
+            except KeyError:
+                pass
+
+    def _get_cache(self):
+        return self.__cache
+
+    def set(self, key, value):
+        with self.lock:
+            try:
+                self.__cache[key] = value
+            except KeyError:
+                pass
+
+    def keys(self):
+        return self.__cache.keys()
+
+    def __len__(self):
+        return len(self.__cache)
+
+    def __iter__(self):
+        return iter(self.__cache)
+
+    def __contains__(self, key):
+        return key in self.__cache
+
+
+class ThreadSafeTTLCache(BaseThreadSafeCache):
+    def create_cache(self, *args, **kwargs):
+        return cachetools.TTLCache(*args, **kwargs)
+
+
+def make_outkey_vin(txhex, vout):
+    txbin = binascii.unhexlify(txhex) if isinstance(txhex, str) else txhex
+    assert isinstance(vout, int)
+
+    tx = bitcoinlib.core.CTransaction.deserialize(txbin)
+    outkey = [(vin.prevout.hash, vin.prevout.n) for vin in tx.vin]
```

Comment:
Is the warning bellow correct?

Reply:
hum.. can't see the warning anymore! @wilfreddenton do you remember it ?